wsproto/connection.py

```diff
@@ -378,8 +378,10 @@ def _process_connection_request(self, event):
         if b'sec-websocket-version' not in headers:
             return ConnectionFailed(CloseReason.PROTOCOL_ERROR,
                                     "Missing Sec-WebSocket-Version header")
-        # XX FIXME: need to check Sec-Websocket-Version, and respond with a
-        # 400 if it's not what we expect
+        versions = _split_comma_header(headers[b'sec-websocket-version'])
+        if WEBSOCKET_VERSION not in versions:
+            return ConnectionFailed(CloseReason.PROTOCOL_ERROR,
+                                    "Unsupported Sec-WebSocket-Version")
```

Comment:
So... this is a little tricky. According to the spec (section 4.2.2), servers that get an unrecognized version MUST respond in a particular way:
```
          The |Sec-WebSocket-Version| header field in the client's
          handshake includes the version of the WebSocket Protocol with
          which the client is attempting to communicate.  If this
          version does not match a version understood by the server, the
          server MUST abort the WebSocket handshake described in this
          section and instead send an appropriate HTTP error code (such
          as 426 Upgrade Required) and a |Sec-WebSocket-Version| header
          field indicating the version(s) the server is capable of
          understanding.
```
Right now we don't really have any way to handle this, right? When a connection fails, we just bail out, and we expect the user to somehow construct an appropriate HTTP response? This seems a bit unfriendly in general, and even more so in this case where we're not even giving the user enough information to construct that response correctly... maybe we should be generating the response ourselves? @jeamland, @lukasa would appreciate thoughts here.

Reply:
You are right - this change simply dumps the problem onto the user to take care off.
The problem for me (in the context of @mitmproxy) is that we want to craft our own response, so wsproto should not respond directly.

I agree it is not perfect, however, still better than simply ignoring the version identifier at all.
Not sure about which version are actually out there, but I am not aware of anything other than version 13...?

Reply:
There are other versions that represent draft versions of the spec, before it was finalized and became an RFC. Some libraries still support them. In the future there could in theory be more versions. But 13 is currently the only standard version, so that's all wsproto supports.

I'm not sure why you want to craft your own response, given that the spec kinda only gives you one option :-). But also, if you want to hack some odd behavior in here, you do always have the option of ignoring the bytes that wsproto tells you to send, and replacing them with your own instead.

Reply:
mitmproxy is a man-in-the-middle web proxy for HTTP 1 & 2, WebSockets, raw TCP,...
so there all sorts of reasons why our users might want to do things 😄 
We also want to to do the handshake ourself, because it is part of our HTTP parsing pipeline already.

Everything else in `_process_connection_request` raises an exception, so I think it is reasonable to do the same for unsupported versions.

How about we introduce a `class UnsupportedVersion(ConnectionFailed)` exception, and provide a `def generate_unsupported_version()` function to return a response with the appropriate data?

Reply:
If you're doing the handshake yourself, then why do you care how wsproto handles the handshake? :-)

> How about we introduce a class `UnsupportedVersion(ConnectionFailed)` exception, and provide a `def generate_unsupported_version()` function to return a response with the appropriate data?

This seems like a lot of boilerplate for users to worry about...

Reply:
So, to summarize and finally get a consensus on this:

You don't like throwing an exception if the version doesn't match what wsproto supports.
wsproto should directly respond with the HTTP response message to indicate this error.

This is what I currently plan to use in mitmproxy:
https://github.com/mitmproxy/mitmproxy/blob/c9b8b2e3a3a3fef18dfa66bcd9288f12bef357b6/mitmproxy/proxy/protocol/websocket.py#L51-L75

In L67 https://github.com/mitmproxy/mitmproxy/blob/c9b8b2e3a3a3fef18dfa66bcd9288f12bef357b6/mitmproxy/proxy/protocol/websocket.py#L67 I need some way of identifying the error case (version mismatch). If wsproto only tells me this via an HTTP response, I have to rely on parsing the output of `bytes_to_send` to check this...

Do you feel more comfortable with a new connection state: `ConnectionState.VersionMismatch` to indicate this error?